src/sim/kernel_workload.cc

```diff
@@ -102,12 +102,29 @@ KernelWorkload::initState()
     auto mapper = [this](Addr a) {
         return (a & _loadAddrMask) + _loadAddrOffset;
     };
+    auto memory_size = system->memSize();
     if (params().object_file != "")  {
         if (params().addr_check) {
+            // Check that memory size is large enough to contain the kernel.
+            if (memory_size < _end){
```

Comment:
My gut feeling is that there might be something wrong here in some cases.... for example what happens if the image starts at an address different than 0? then you might end up with an _end bigger than memory size even if you have allocated enough memory..

Reply:
I think this is fair, would it be reasonable to quickly determine the image size somewhere (taking into account start and end --- we can assume kernel memory is contiguous?). Then down in the fatal-if we do:

```cpp
!system->isMemAddr(mapper(_end)),
    "Kernel is mapped to invalid location (not memory). " +
    kernel_size > system->memSize() ? " The kernel size (" +   
    kenel_size + " blocks) appears to be larger than total system memory " + system->memSize() + " blocks). "
    "Try increasing system memory" : ""
```

(I'm sure this isn't real CPP so please consider if Pseudocode).

The rest of the error we can remove and keep the error in one line.

Reply:
This looks like a good solution to me!